plugin/evm/vm.go

```diff
@@ -1345,3 +1398,99 @@ func (vm *VM) createDatabase(dbConfig avalancheNode.DatabaseConfig) (database.Da
 
 	return db, nil
 }
+
+func (vm *VM) Connected(ctx context.Context, nodeID ids.NodeID, version *version.Application) error {
+	if err := vm.uptimeManager.Connect(nodeID); err != nil {
+		return fmt.Errorf("uptime manager failed to connect node %s: %w", nodeID, err)
+	}
+	return vm.Network.Connected(ctx, nodeID, version)
+}
+
+func (vm *VM) Disconnected(ctx context.Context, nodeID ids.NodeID) error {
+	if err := vm.uptimeManager.Disconnect(nodeID); err != nil {
+		return fmt.Errorf("uptime manager failed to disconnect node %s: %w", nodeID, err)
+	}
+
+	return vm.Network.Disconnected(ctx, nodeID)
+}
+
+func (vm *VM) dispatchUpdateValidators(ctx context.Context) {
+	ticker := time.NewTicker(loadValidatorsFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			vm.ctx.Lock.Lock()
+			if err := vm.performValidatorUpdate(ctx); err != nil {
+				log.Error("failed to update validators", "error", err)
+			}
+			vm.ctx.Lock.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// performValidatorUpdate updates the validator state with the current validator set
+// and writes the state to the database.
+func (vm *VM) performValidatorUpdate(ctx context.Context) error {
+	now := time.Now()
+	log.Debug("performing validator update")
+	// get current validator set
+	currentValidatorSet, _, _, err := vm.ctx.ValidatorState.GetCurrentValidatorSet(ctx, vm.ctx.SubnetID)
+	if err != nil {
+		return fmt.Errorf("failed to get current validator set: %w", err)
+	}
+
+	log.Info("updating validators", "validatorSet", currentValidatorSet)
+
+	// load the current validator set into the validator state
+	if err := loadValidators(vm.validatorState, currentValidatorSet); err != nil {
+		return fmt.Errorf("failed to load current validators: %w", err)
+	}
+
+	// write validators to the database
+	if err := vm.validatorState.WriteState(); err != nil {
+		return fmt.Errorf("failed to write validator state: %w", err)
+	}
+
+	// TODO: add metrics
+	log.Debug("validator update complete", "duration", time.Since(now))
+	return nil
+}
+
+// TODO: cache the last updated height and then load if needed
```

Comment:
not sure if this would bring us any viable advantage, since P-chain height will likely be changing every loadValidatorsFrequency (1 minute).

Reply:
I think we can remove this TODO